ts/receiver/multidevice.ts

```diff
@@ -63,8 +73,363 @@ export async function handleUnpairRequest(
+  envelope: EnvelopePlus,
+  pairingAuthorisation: SignalService.IPairingAuthorisationMessage,
+  dataMessage: SignalService.IDataMessage,
+  syncMessage: SignalService.ISyncMessage
+) {
+  const { ConversationController, libloki, Whisper } = window;
+
+  const valid = await libloki.crypto.validateAuthorisation(
+    pairingAuthorisation
+  );
+  const alreadySecondaryDevice = !!window.storage.get('isSecondaryDevice');
+  let removedFromCache = false;
+  if (alreadySecondaryDevice) {
+    window.log.warn(
+      'Received an unexpected pairing authorisation (device is already paired as secondary device). Ignoring.'
+    );
+  } else if (!valid) {
+    window.log.warn(
+      'Received invalid pairing authorisation for self. Could not verify signature. Ignoring.'
+    );
+  } else {
+    const { primaryDevicePubKey, grantSignature } = pairingAuthorisation;
+    if (grantSignature) {
+      // Authorisation received to become a secondary device
+      window.log.info(
+        `Received pairing authorisation from ${primaryDevicePubKey}`
+      );
+      // Set current device as secondary.
+      // This will ensure the authorisation is sent
+      // along with each session request.
+      window.storage.remove('secondaryDeviceStatus');
+      window.storage.put('isSecondaryDevice', true);
+      window.storage.put('primaryDevicePubKey', primaryDevicePubKey);
+      await MultiDeviceProtocol.savePairingAuthorisation(
+        pairingAuthorisation as Data.PairingAuthorisation
+      );
+      const primaryConversation = await ConversationController.getOrCreateAndWait(
+        primaryDevicePubKey,
+        'private'
+      );
+      primaryConversation.trigger('change');
+      Whisper.events.trigger('secondaryDeviceRegistration');
+      // Update profile
+      if (dataMessage) {
+        const { profile, profileKey } = dataMessage;
+
+        if (profile && profileKey) {
+          const ourNumber = window.storage.get('primaryDevicePubKey');
+          const me = window.ConversationController.get(ourNumber);
+          if (me) {
+            await updateProfile(me, profile, profileKey);
+          }
+        } else {
+          window.log.warn('profile or profileKey are missing in DataMessage');
+        }
+      }
+      // Update contact list
+      if (syncMessage && syncMessage.contacts) {
+        // Note: we do not return here because we don't want to block the next message on
+        //   this attachment download and a lot of processing of that attachment.
+        // This call already removes the envelope from the cache
+        void handleContacts(envelope, syncMessage.contacts);
+        removedFromCache = true;
+      }
+    } else {
+      window.log.warn('Unimplemented pairing authorisation message type');
+    }
+  }
+  if (!removedFromCache) {
+    await removeFromCache(envelope);
+  }
+}
+
+function parseContacts(arrbuf: ArrayBuffer): Array<any> {
+  const buffer = new window.dcodeIO.ByteBuffer();
```

Comment:
We can again import `ByteBuffer` directly

Reply:
updated